src/Data/ProgrammesDb/EntityRepository/CoreEntityRepository.php

```diff
@@ -203,6 +203,38 @@ public function findByIds(array $ids): array
             ->getQuery()->getResult(Query::HYDRATE_ARRAY);
     }
 
+    /**
+     * @param int[]    $ancestryDbIds
+     * @param string   $entityType
+     * @param int|null $limit
+     * @param int      $offset
+     * @return array
+     */
+    public function findProgrammesByAncestryAndType(array $ancestryDbIds, string $entityType, ?int $limit, int $offset) : array
+    {
+        $this->assertEntityType($entityType, ['Clip', 'Episode', 'Series']);
+
+        $qb = $this->getEntityManager()->createQueryBuilder()
+            ->addSelect(['programme', 'masterBrand', 'image', 'mbImage'])
+            ->from('ProgrammesPagesService:' . $entityType, 'programme')
+            ->leftJoin('programme.image', 'image')
+            ->leftJoin('programme.masterBrand', 'masterBrand')
+            ->leftJoin('masterBrand.image', 'mbImage')
+            ->andWhere('programme.ancestry LIKE :ancestry')
+            ->andWhere('programme.streamable = 1')
+            ->setFirstResult($offset)
+            ->setMaxResults($limit)
+            ->setParameter('ancestry', $this->ancestryIdsToString($ancestryDbIds) . '%');
+
+        if (in_array($entityType, ['Episode', 'Clip'])) {
+            // Doctrine will complain loudly if the entity does not have the
+            // field to order by on it.
+            $qb->orderBy('programme.streamableFrom', 'DESC');
+        }
+
+        return $qb->getQuery()->getResult(Query::HYDRATE_ARRAY);
```

Comment:
You'll need to resolve the ancestry here otherwise we won't be able to build the programme objects later.